awx/main/models/events.py

```diff
@@ -0,0 +1,685 @@
+    )
+    uuid = models.CharField(
+        max_length=1024,
+        default='',
+        editable=False,
+    )
+    counter = models.PositiveIntegerField(
+        default=0,
+        editable=False,
+    )
+    stdout = models.TextField(
+        default='',
+        editable=False,
+    )
+    verbosity = models.PositiveIntegerField(
+        default=0,
+        editable=False,
+    )
+    start_line = models.PositiveIntegerField(
+        default=0,
+        editable=False,
+    )
+    end_line = models.PositiveIntegerField(
+        default=0,
+        editable=False,
+    )
+
+    def get_absolute_url(self, request=None):
+        return reverse('api:ad_hoc_command_event_detail', kwargs={'pk': self.pk}, request=request)
+
+    def __unicode__(self):
+        return u'%s @ %s' % (self.get_event_display(), self.created.isoformat())
+
+    def save(self, *args, **kwargs):
+        # If update_fields has been specified, add our field names to it,
+        # if it hasn't been specified, then we're just doing a normal save.
+        update_fields = kwargs.get('update_fields', [])
+        res = self.event_data.get('res', None)
+        if self.event in self.FAILED_EVENTS:
+            if not self.event_data.get('ignore_errors', False):
+                self.failed = True
+                if 'failed' not in update_fields:
+                    update_fields.append('failed')
+        if isinstance(res, dict) and res.get('changed', False):
+            self.changed = True
+            if 'changed' not in update_fields:
+                update_fields.append('changed')
+        super(BaseCommandEvent, self).save(*args, **kwargs)
+
+    @classmethod
+    def create_from_data(self, **kwargs):
+        # Convert the datetime for the ad hoc command event's creation
+        # appropriately, and include a time zone for it.
+        #
+        # In the event of any issue, throw it out, and Django will just save
+        # the current time.
+        try:
+            if not isinstance(kwargs['created'], datetime.datetime):
+                kwargs['created'] = parse_datetime(kwargs['created'])
+            if not kwargs['created'].tzinfo:
+                kwargs['created'] = kwargs['created'].replace(tzinfo=utc)
+        except (KeyError, ValueError):
+            kwargs.pop('created', None)
+
+        # events that not jobs/adhoc (generic) are just event=verbose
+        if 'job_id' not in kwargs and 'ad_hoc_command_id' not in kwargs:
+            kwargs['event'] = 'verbose'
+
+        # Sanity check: Don't honor keys that we don't recognize.
+        valid_keys = {'ad_hoc_command_id', 'unified_job_id', 'event',
+                      'event_data', 'created', 'counter', 'uuid', 'stdout',
+                      'start_line', 'end_line', 'verbosity'}
+        for key in kwargs.keys():
+            if key not in valid_keys:
+                kwargs.pop(key)
+
+        return self.objects.create(**kwargs)
+
+
+class AdHocCommandEvent(BaseCommandEvent):
```

Comment:
I extracted the shared job event functionality into a shared abstract base class `BaseCommandEvent`, which `AdhocCommandEvent` and `GenericCommandEvent` share.  The adhoc version has some extras (like the `ad_hoc_command` and it's indexes), and references to the host/hostname.

Reply:
You could probably make then same argument for `failed` and `changed` (they're always `False`).

Reply:
Let me just link #364 in here. True, even _if_ we get that in, that won't assign a changed status to the individual events from inventory updates... but why not? Maybe it could happen eventually.